### PR TITLE
Add support for deflate compression

### PR DESCRIPTION
**Pull request by szarsti**

### CODE REVIEW

File: src/avro_ocf.erl [hunk 1]
```diff
@@ -190,25 +203,29 @@ decode_stream(SchemaStore, Type, Bin) when is_binary(Bin) ->
...
 %% @private
--spec decode_block(schema_store(), avro_type(),
+-spec decode_block(schema_store(), avro_type(), avro_codec(),
                    binary(), integer(), [avro:out()]) -> [avro:out()].
-decode_block(_Store, _Type, <<>>, 0, Acc) -> Acc;
-decode_block(Store, Type, Bin, Count, Acc) ->
+decode_block(_Store, _Type, _Codec, <<>>, 0, Acc) -> Acc;
+decode_block(Store, Type, deflate, Bin, Count, Acc) ->
+  Decompressed = zlib:unzip(Bin),
+  io:format("==> Decompressed: ~p~n", [Decompressed]),
```
Comment (id) [inline src/avro_ocf.erl]: debug info?

File: src/avro_ocf.erl [hunk 2]
```diff
@@ -236,6 +253,27 @@ init_schema_store(Schema) ->
...
+    {_, <<"null">>} ->
+      null;
+    {_, <<"deflate">>} ->
+      deflate
+  end.
+
+%% @private Encode block according to selected codec
+-spec encode_block([{binary(), binary()}], iolist()) -> binary().
+encode_block(Meta, Data) ->
+  io:format("Using codec: ~p~n",  [get_codec(Meta)]),
```
Comment (id) [inline src/avro_ocf.erl]: debug info?

**Comment by szarsti** [comment]
Sorry for missing my debugging printouts. I have cleaned them now.

File: src/avro_ocf.erl [hunk 3]
```diff
@@ -146,12 +145,15 @@ make_header(Type) ->
 -spec make_header(avro_type(), extra_meta()) -> header().
 make_header(Type, ExtraMeta0) ->
   ExtraMeta = validate_extra_meta(ExtraMeta0),
```
Comment (zmstone) [inline src/avro_ocf.erl]: this call will crash (at line 169) if `<<"avro.codec">>` is given in ExtraMeta0

Comment (zmstone) [inline src/avro_ocf.erl]: sorry, missed line 176.

Comment (zmstone) [inline src/avro_ocf.erl]: The name `ExtraMeta` now is a bit confusing.
Would like to have it renamed to just 'Meta', and spec renamed to `meta()`.

Also, please update the docs for API functions (`make_header/2` and `write_file/5`)
to have compression option explained

File: src/avro_ocf.erl [hunk 4]
```diff
@@ -146,12 +145,15 @@ make_header(Type) ->
...
+  Meta = case lists:keyfind(<<"avro.codec">>, 1, ExtraMeta) of
+           false ->
```
Comment (zmstone) [inline src/avro_ocf.erl]: please remove the trailing white-space

File: test/avro_ocf_tests.erl [hunk 5]
```diff
@@ -59,6 +68,19 @@ write_file_test() ->
...
+  Type = avro_record:type("rec", Fields, [{namespace, "defalate.ocf.test"}]),
+  Obj = [{"f1", 1}, {"f2", "foo"}],
+  Meta = [{<<"avro.codec">>, <<"deflate">>}],
+  ok = avro_ocf:write_file(OcfFile, Store, Type, [Obj], Meta),
+  {_Header, Type, Objs} = avro_ocf:decode_file(OcfFile),
+  ?assertEqual([[{<<"f1">>, 1}, {<<"f2">>, <<"foo">>}]], Objs).
+
 root_level_union_test() ->
   OcfFile = filename:join([priv_dir(), "union.ocf.test"]),
   Store = undefined, %% should not require lookup
```
Comment (zmstone) [inline test/avro_ocf_tests.erl]: We have 100% line coverage for `erlavro`, please make sure all lines are covered by tests.
currently below lines are not covered:
```
================ avro_ocf ================
182: is_invalid_codec_meta(<<"avro.codec">>, <<"null">>) -> false;
184: is_invalid_codec_meta(<<"avro.codec">>, _) -> true;
```

check the lines by:
```
make test
./scripts/cover-print-not-covered-lines.escript .eunit/cover.coverdata
```

File: src/avro_ocf.erl (hunk 3 above)
Comment (szarsti) [inline src/avro_ocf.erl]: OK, that sounds reasonable. I'll prepare fix with your suggestions tomorrow morning.

**Comment by szarsti** [comment]
I think that solves all your comments @zmstone. Please let me know if you would like me to anything more.

File: src/avro_ocf.erl [hunk 6]
```diff
@@ -90,11 +89,12 @@ decode_binary(Bin) ->
 write_file(Filename, SchemaStore, Schema, Objects) ->
   write_file(Filename, SchemaStore, Schema, Objects, []).
 
-%% @doc Write objects in a single block to the given file name.
+%% @doc Write objects in a single block to the given file name with custom
+%% metadata. @see make_header/2 for details about use of meta data.
```
Comment (zmstone) [inline src/avro_ocf.erl]: `@end`

File: src/avro_ocf.erl [hunk 7]
```diff
@@ -142,37 +142,54 @@ append_file(Fd, Header, SchemaStore, Schema, Objects) ->
...
-%% @doc Make ocf header, and append the given extra metadata fields.
--spec make_header(avro_type(), extra_meta()) -> header().
-make_header(Type, ExtraMeta0) ->
-  ExtraMeta = validate_extra_meta(ExtraMeta0),
+%% @doc Make ocf header, and append the given metadata fields.
+%% You can use &gt;&gt;"avro.codec"&lt;&lt; metadata field to choose what data
+%% block coding should be used. Supported values are `&gt;&gt;"null"&lt;&lt;'
+%% (default) and `&gt;&gt;"deflate"&lt;&lt;' (compressed). Other values in
+%% `avro' namespace are reserved for internal use and can't be set.
+%% Other than that you are free to provide any custom metadata.
```
Comment (zmstone) [inline src/avro_ocf.erl]: `@end`

File: src/avro_ocf.erl [hunk 8]
```diff
@@ -142,37 +142,54 @@ append_file(Fd, Header, SchemaStore, Schema, Objects) ->
 make_header(Type) ->
   make_header(Type, _ExtraMeta = []).
 
-%% @doc Make ocf header, and append the given extra metadata fields.
--spec make_header(avro_type(), extra_meta()) -> header().
-make_header(Type, ExtraMeta0) ->
-  ExtraMeta = validate_extra_meta(ExtraMeta0),
+%% @doc Make ocf header, and append the given metadata fields.
+%% You can use &gt;&gt;"avro.codec"&lt;&lt; metadata field to choose what data
```
Comment (zmstone) [inline src/avro_ocf.erl]: quote code in ``'`. e.g. `<<"null">>'
this is more clear to read in code editor.
and will be translated to `<code>&lt;&lt;"null"&gt;&gt;</code>` in html.

**Comment by zmstone** [comment]
@szarsti please bump version in .app.src to `2.2.0`

File: src/avro_ocf.erl (hunk 8 above)
Comment (szarsti) [inline src/avro_ocf.erl]: Thanks for hints about edoc formatting

**Comment by zmstone** [comment]
Thanks for your contribution. @szarsti

**Pull request merged**